### PR TITLE
KAFKA-6958: Overload KStream methods to allow to name operation name  using the new Named class

### PR DESCRIPTION
**Pull request by fhussonnois**
Hi @mjsax @bbejeck

This is the 3rd PR for the KIP-307.

NOTE : [PR 6410](https://github.com/apache/kafka/pull/6410) should be merge first

Thanks a lot for the review.

### CODE REVIEW

**Review by bbejeck — COMMENTED** [review]
Overall this PR looks good to me, as any changes I'd suggest have already been addressed ion #6409 and #6410 this PR seems to be focused on adding the `Named` parameter overload to `KStream` and `KStreamImpl` and updating the tests.

**Comment by bbejeck** [comment]
@fhussonnois please rebase this PR

**Comment by bbejeck** [comment]
ping @vvcephei and @ableegoldman for reviews

**Comment by bbejeck** [comment]
Hi @fhussonnois #6410 was just merged, so if you could rebase this PR, we can make another pass and get this one merged as well.

Thanks for your work and patience!

**Comment by fhussonnois** [comment]
@bbejeck This PR has been rebase. 

The KIP has been updated to include the newest methods that was missing : 
`flatTransformValues(ValueTransformerWithKeySupplier, Named,  String... )`
`flatTransformValues(ValueTransformerSupplier, Named, String...)`

**Comment by bbejeck** [comment]
Thanks @fhussonnois! taking a look at this today.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: nit: add `{@link Named}` here and elsewhere below

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:161]: Why not a `requireNonNull` here as well for `Named`?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: nit: should be `named` can't be null

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: nit: add opening and closing curly braces for all the `if` / `else` blocks on lines 1102 to 1108

**Review by bbejeck — COMMENTED** [review]
Thanks @fhussonnois just some very minor comments, otherwise LGTM

**Comment by bbejeck** [comment]
call for second review any of @guozhangwang , @mjsax @vvcephei, @ableegoldman, @abbccdda

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:482]: Do we consider refactoring these tests with one comment function? I could see L485-487 are repeated multiple times.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]:
```suggestion
     * Furthermore, via {@link org.apache.kafka.streams.processor.Punctuator#punctuate()} the processing progress can
```

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: Please also fix this on the original method (line 1739)

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]:
```suggestion
     * @param named     a {@link Named} config used to name the processor in the topology.
```

Or something similar, it just reads a little bit on the terse side, which I think would confuse people initially. Please also apply this feedback to all the other param docs.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]:
```suggestion
     * If the return value of {@link ValueTransformer#transform(Object) ValueTransformer#transform()} is {@code null}, no
```

Please also fix this on the original method

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]:
```suggestion
     * is {@code null}, no records are emitted.
```

Please also fix this on the original method

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]:
```suggestion
     * {@link java.lang.Iterable Iterable} or {@code null}, no records are emitted.
```

Please also fix this on the original method

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]:
```suggestion
     * is an empty {@link java.lang.Iterable Iterable} or {@code null}, no records are emitted.
```

Please also fix this on the original method

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]:
```suggestion
     * If the return value of {@link ValueTransformer#transform(Object) ValueTransformer#transform()} is {@code null}, no
```

Please also fix this on the original method

File: streams/src/main/java/org/apache/kafka/streams/kstream/Named.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Named.java]: Do we need a copy constructor here? It seems purely in service of the NamedInternal constructor, but that class could trivially just use the main constructor. All in all, it seems like we're adding extra hops to follow while reading the code, without much benefit.

Also, can we move the requireNotNull inside the main constructor and remove it from the other methods in this class, while also making the field final? That's the way to guarantee that the field is never null when the name is set.

Also, having `empty` with a nullable name, and then having an extra method later to generate the name if needed seems unnecessarily complex. All processor nodes ultimately need names. Can we just make the name field never-null, and replace `empty()` with `generated(prefix)` to just make the name immediately?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: Can we just inline `doFlatTransform` now? There's no need to have 3 layers of indirection to build these processors.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]:
```suggestion
        builder.addGraphNode(streamsGraphNode, transformNode);
```

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]:
```suggestion
        final JoinedInternal<K, V, VO> joinedInternal = new JoinedInternal<>(joined);
```

File: streams/src/test/java/org/apache/kafka/streams/kstream/RepartitionTopicNamingTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/RepartitionTopicNamingTest.java:569]: Any idea why these processors changed order? It could indicate a deeper problem.

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:530]:
```suggestion
    @SuppressWarnings({"unchecked", "rawtypes"})
```

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]:
```suggestion
    @SuppressWarnings({"unchecked", "rawtypes"})
```

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]: This variable is unused. Is there some assertion we can make on it?

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]:
```suggestion
    @SuppressWarnings({"unchecked", "rawtypes"})
```

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java]:
```suggestion
    @SuppressWarnings({"unchecked", "rawtypes"})
```

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:707]: (applies to the methods below)

Just as cleanup, can you:
* add `@SuppressWarnings("rawtypes")` to `assertSpecifiedNameForOperation`?
* make `assertSpecifiedNameForOperation` and `assertSpecifiedNameForStateStore` to be `static`?
* remove unnecessary qualifier `Assert.` from `assertEquals`?

Thanks!

**Review by vvcephei — CHANGES_REQUESTED** [review]
Thanks for the PR @fhussonnois !

According to my code analyser, several of the new methods are unused. Can we make sure we have test coverage verifying the new overloads work properly? I know this may seem paranoid, but we've had several embarrasing bugs in the Scala API, which should be similarly trivial, due to typos in the implementation code that could have been caught if they had any test coverage at all.

I had some thoughts about Named/NamedInternal, which I made inline.

There were several nits, which I just sent "suggestions" for.

Not sure if it helps, but I created a child page to your KIP, so the reviewers can strike though methods as you send PRs for them. (I just did the methods in this PR). I found it useful to go down the checklist for this PR.

Looks good, thanks!

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: @fhussonnois can we move the logic from here to the `NamedInternal` to always generate the name to ensure we increment the counter for backward compatibility?  

Something along the lines of 
```java
private String orElseGet(final Supplier<String> supplier) {
        final String generatedName = supplier.get();
        return Optional.ofNullable(this.name).orElseGet(() -> generatedName);
    }
```

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: Same for here

File: streams/src/test/java/org/apache/kafka/streams/kstream/RepartitionTopicNamingTest.java
Comment (fhussonnois) [reply streams/src/test/java/org/apache/kafka/streams/kstream/RepartitionTopicNamingTest.java:569]: Not sure but I think, the reason is because nodes are sorted : https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java#L761

**Comment by fhussonnois** [comment]
@bbejeck This PR has been rebased, most of the feedbacks has been resolved.

**Comment by bbejeck** [comment]
both Java 8 and Java 11 failed, test results already removed.

retest this please

**Comment by fhussonnois** [comment]
@bbejeck if everything is OK we this PR, could we merge it to move forward to the next one ?

**Review by bbejeck — APPROVED** [review]
Thanks @fhussonnois LGTM modulo comments left over from @vvcephei and @abbccdda

**Comment by bbejeck** [comment]
Hi @fhussonnois if you can rebase this PR we can then merge it, then move on to parts 4 and 5.

Thanks!

**Comment by fhussonnois** [comment]
Hi @bbejeck , this PR has been rebased.

**Comment by bbejeck** [comment]
both Java 8 and Java 11 failed, test results already cleared out.

retest this please

**Comment by bbejeck** [comment]
Java 8 passed Java 11 failed with `kafka.integration.UncleanLeaderElectionTest.testUncleanLeaderElectionDisabled`

retest this please

**Comment by bbejeck** [comment]
Java 8 failed, Java 11 passed

retest this please

**Comment by bbejeck** [comment]
retest this please

**Comment by bbejeck** [comment]
Merged #6411 into trunk.

**Comment by bbejeck** [comment]
Thank you @fhussonnois!

**Pull request merged**
